pkg/networkserver/redis/application_uplink_queue.go

```diff
@@ -61,9 +69,26 @@ func (q *ApplicationUplinkQueue) Add(ctx context.Context, ups ...*ttnpb.Applicat
 			return err
 		}
 
+		var (
+			streamID             string
+			maxLen, maxLenApprox int64
+		)
+		switch up.Up.(type) {
+		case *ttnpb.ApplicationUp_JoinAccept:
+			streamID = q.uidJoinAcceptKey(uid)
+			maxLenApprox = q.maxLen
+		case *ttnpb.ApplicationUp_DownlinkQueueInvalidated:
+			streamID = q.uidInvalidationKey(uid)
+			maxLen = 1
+		default:
+			streamID = q.uidGenericUplinkKey(uid)
+			maxLenApprox = q.maxLen
+		}
+
 		if err = q.redis.XAdd(&redis.XAddArgs{
-			Stream:       q.uidUplinkKey(uid),
-			MaxLenApprox: q.maxLen,
+			Stream:       streamID,
+			MaxLen:       maxLen,
```

Comment:
Is it valid to specify both ? As far as I know the command syntax is to add `~` when you want to approximate, so how would specifying both work ?

Reply:
One of them is always zero